### ISSUE TITLE
nothing in `(x else nothing)()` is cast to raw Callable, wrong return type

### ISSUE THREAD

**Issue by lucaswerkmeister** [ceylon/ceylon-compiler#2065]
When `nothing` appears in an `else` expression, which is then called, it compiles to something like this:
```java
(
    let
    {
        .ceylon.language.Callable<? extends XYZ> $ceylontmp$0 = .tmp.f1_.get_();
    }
    returning $ceylontmp$0 != null ? $ceylontmp$0 : (.ceylon.language.Callable).ceylon.language.nothing_.get_();
).$call$()
```
Note that the temporary value has a correctly parametrized type, but `nothing` is cast to raw `Callable`. Because of this, `.$call$()` returns `Object` instead of `XYZ`.

This can result in various backend errors, for instance:
```ceylon
Empty()? f1 = null;
Empty c1 = (f1 else nothing)();
```
> incompatible types
  required: `ceylon.language.Empty`
  found:    `java.lang.Object`

Warning: This code works for a few special types, like `Null` (erases to `Object`) or `String` (calls `.toString()` on the result).

```ceylon
Anything()()? f2 = null;
Anything c2 = (f2 else nothing)()();
```
> error: cannot find symbol
  symbol:   method `$call$()`
  location: class `Object`

**Comment by tombentley** [comment]
Fixed by the same changes which fixed #2066.

**Comment by lucaswerkmeister** [comment]
Great, thanks!
